File: api-automation-java/src/test/java/com/orangehrm/api/tests/EmployeeWorkflowTests.java
package com.orangehrm.api.tests;

import com.orangehrm.api.support.ApiTestBase;
import com.orangehrm.api.support.AuthClient;
import com.orangehrm.api.support.OrangeHrmApi;
import com.orangehrm.api.support.TestConfig;
import io.restassured.http.Cookies;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.hamcrest.Matchers.*;

/**
 * Employee-related workflows and backend business-logic validations.
 *
 * Note:
 * OrangeHRM employee creation endpoints can require specific fields and permissions.
 * This sample uses "Admin - Job Titles" as a stable CRUD-like workflow, and includes
 * generic collection paging validations which are business-logic adjacent.
 */
public class EmployeeWorkflowTests extends ApiTestBase {

    @Test
    @DisplayName("Create -> Delete workflow (Admin: Job Titles) should succeed")
    void createAndDeleteJobTitle() {
        Cookies cookies = AuthClient.loginWithCookies(TestConfig.adminUsername(), TestConfig.adminPassword());
        String endpoint = OrangeHrmApi.v2("/admin/job-titles");

        // Create
        int createdId =
                OrangeHrmApi.authed(cookies)
                        .body(Map.of(
                                "title", "API-AUTO-" + System.currentTimeMillis(),
                                "description", "Created by API automation suite",
                                "specification", null,
                                "note", "cleanup expected"
                        ))
                        .when()
                        .post(endpoint)
                        .then()
                        .statusCode(200)
                        .body("data.id", notNullValue())
                        .extract()
                        .path("data.id");

        // Verify list contains created record (basic business logic validation)
        OrangeHrmApi.authed(cookies)
                .queryParam("limit", 50)
                .queryParam("offset", 0)
                .when()
                .get(endpoint)
                .then()
                .statusCode(200)
                .body("data", notNullValue())
                .body("meta", notNullValue());

        // Delete
        OrangeHrmApi.authed(cookies)
                .body(Map.of("ids", new int[]{createdId}))
                .when()
                .delete(endpoint)
                .then()
                .statusCode(200)
                .body("data", hasItem(createdId));
    }

    @Test
    @DisplayName("Pagination business logic: limit/offset should be honored (shape + no errors)")
    void paginationValidation() {
        Cookies cookies = AuthClient.loginWithCookies(TestConfig.adminUsername(), TestConfig.adminPassword());
        String endpoint = OrangeHrmApi.v2("/admin/job-titles");

        OrangeHrmApi.authed(cookies)
                .queryParam("limit", 1)
                .queryParam("offset", 0)
                .when()
                .get(endpoint)
                .then()
                .statusCode(200)
                .body("data", notNullValue())
                // Many collections return <= limit
                .body("data.size()", lessThanOrEqualTo(1))
                .body("meta", notNullValue());

        OrangeHrmApi.authed(cookies)
                .queryParam("limit", 1)
                .queryParam("offset", 1)
                .when()
                .get(endpoint)
                .then()
                .statusCode(200)
                .body("data", notNullValue())
                .body("data.size()", lessThanOrEqualTo(1))
                .body("meta", notNullValue());
    }

    @Test
    @DisplayName("Validation business logic: deleting with empty ids should be rejected")
    void deleteValidationShouldFailOnEmptyIds() {
        Cookies cookies = AuthClient.loginWithCookies(TestConfig.adminUsername(), TestConfig.adminPassword());
        String endpoint = OrangeHrmApi.v2("/admin/job-titles");

        OrangeHrmApi.authed(cookies)
                .body(Map.of("ids", new int[]{}))
                .when()
                .delete(endpoint)
                .then()
                .statusCode(anyOf(is(400), is(422)))
                .body(anything());
    }
}

File: api-automation-csharp/OrangeHrm.ApiTests/Tests/EmployeeWorkflowTests.cs
using System;
using System.Net;
using System.Text.Json;
using System.Threading.Tasks;
using NUnit.Framework;
using RestSharp;
using OrangeHrm.ApiTests.Support;

namespace OrangeHrm.ApiTests.Tests;

[TestFixture]
public class EmployeeWorkflowTests
{
    [Test]
    public async Task Create_And_Delete_JobTitle_Workflow_Should_Succeed()
    {
        var client = new OrangeHrmClient();
        await client.LoginWithCookiesAsync(TestConfig.AdminUsername, TestConfig.AdminPassword);

        var endpoint = OrangeHrmClient.V2("/admin/job-titles");

        // Create
        var createReq = new RestRequest(endpoint, Method.Post);
        createReq.AddJsonBody(new
        {
            title = $"API-AUTO-{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}",
            description = "Created by API automation suite",
            specification = (string?)null,
            note = "cleanup expected"
        });

        var createRes = await client.ExecuteAuthedAsync(createReq);
        Assert.That(createRes.StatusCode, Is.EqualTo(HttpStatusCode.OK));
        Assert.That(createRes.Content, Is.Not.Null.And.Not.Empty);

        using var createJson = OrangeHrmClient.ParseJson(createRes.Content);
        var id = createJson.RootElement.GetProperty("data").GetProperty("id").GetInt32();
        Assert.That(id, Is.GreaterThan(0));

        // Delete
        var deleteReq = new RestRequest(endpoint, Method.Delete);
        deleteReq.AddJsonBody(new { ids = new[] { id } });

        var deleteRes = await client.ExecuteAuthedAsync(deleteReq);
        Assert.That(deleteRes.StatusCode, Is.EqualTo(HttpStatusCode.OK));

        using var deleteJson = OrangeHrmClient.ParseJson(deleteRes.Content);
        var dataArray = deleteJson.RootElement.GetProperty("data");
        Assert.That(dataArray.ValueKind, Is.EqualTo(JsonValueKind.Array));
    }

    [Test]
    public async Task Pagination_Should_Return_At_Most_Limit_Items()
    {
        var client = new OrangeHrmClient();
        await client.LoginWithCookiesAsync(TestConfig.AdminUsername, TestConfig.AdminPassword);

        var endpoint = OrangeHrmClient.V2("/admin/job-titles");

        var req = new RestRequest(endpoint, Method.Get);
        req.AddQueryParameter("limit", "1");
        req.AddQueryParameter("offset", "0");

        var res = await client.ExecuteAuthedAsync(req);
        Assert.That(res.StatusCode, Is.EqualTo(HttpStatusCode.OK));

        using var json = OrangeHrmClient.ParseJson(res.Content);
        var data = json.RootElement.GetProperty("data");
        Assert.That(data.ValueKind, Is.EqualTo(JsonValueKind.Array));
        Assert.That(data.GetArrayLength(), Is.LessThanOrEqualTo(1));
    }

    [Test]
    public async Task Delete_With_Empty_Ids_Should_Be_Rejected()
    {
        var client = new OrangeHrmClient();
        await client.LoginWithCookiesAsync(TestConfig.AdminUsername, TestConfig.AdminPassword);

        var endpoint = OrangeHrmClient.V2("/admin/job-titles");

        var req = new RestRequest(endpoint, Method.Delete);
        req.AddJsonBody(new { ids = Array.Empty<int>() });

        var res = await client.ExecuteAuthedAsync(req);

        Assert.That(res.StatusCode, Is.AnyOf(HttpStatusCode.BadRequest, (HttpStatusCode)422));
    }
}
